QccPack-0.61-1/src/libQccPackWAV/utils/man1/imgdwt.1
.TH IMGDWT 1 "QCCPACK" ""
.SH NAME
imgdwt, imgidwt \- DWT and SA-DWT of an image, and corresponding inverse
transform
.SH SYNOPSIS
.B imgdwt
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-pw
.IR  pcpfile "\|]"
.RB "[\|" \-m
.IR  mask "\|]"
.RB "[\|" \-nl
.IR  levels "\|]"
.I imgfile
.I subband_pyramid
.LP
.B imgidwt
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-pw
.IR  pcpfile "\|]"
.RB "[\|" \-m
.IR  mask "\|]"
.RB "[\|" \-l
.IR  last_subband "\|]"
.I subband_pyramid
.I imgfile
.SH OPTIONS
.TP
.BI \-w " wavelet"
String. 
Name of the wavelet file
.RB ( FBK
or
.B LFT
format); default = "CohenDaubechiesFeauveau.9-7.lft"
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP
.BI \-pw " pcpfile"
String. Name of file containing perceptual weights.
.TP
.BI \-m " mask"
String. Name of file containing the transparency mask for
shape-adaptive transform.
.TP 
.BI \-nl " levels"
Integer. Number of levels of dyadic decomposition; default value is 1 level.
.SH DESCRIPTION
.LP
.B imgdwt
implements a 2D separable discrete wavelet transform (DWT)
of grayscale image
.IR imgfile .
This wavelet decomposition is accomplished by applying a 1D
DWT, first along every row and then along every column.
The output of this process is
a low-resolution approximation (the baseband image)
to the original image, as well as highpass subbands
in the horizontal, vertical, and diagonal directions.
The baseband and the highpass subbands are each
one quarter the size of the original image.
If
.I levels
is greater than 1, then this process is repeated on the baseband;
i.e., the baseband subband is recursively decomposed.
The resulting subbands of these decompositions are recursively nested in the
upper-left of
.IR subband_pyramid 
in a so-called dyadic, or octave, subband-pyramid structure.
.I subband_pyramid
has the same number of rows and columns as
.IR imgfile .
.LP
.I wavelet
is the name of a file containing either the
filter bank 
.RB ( FBK 
format) or lifting scheme
.RB ( LFT
format) to use in performing the DWT.
The name of the filter-bank/lifting-scheme file is given by
.IR wavelet ;
this file must be found in the path given by the
environment variable
.BR QCCPACK_WAVELET_PATH .
The appropriate transform implementation
(filter bank, lifting) is used based upon the type
.RB ( FBK , " LFT" )
of the found file.
If the user does not specify a wavelet,
the default wavelet is a lifting implementation of the
Cohen-Feauveau-Daubechies 9/7 spline with less dissimilar lengths, a
biorthogonal wavelet popular wavelet for image-processing
applications (see Cohen 
.IR "et al" .,
Antonini 
.IR "et al" .).
.LP
.I boundary
gives the method for handling the image boundaries.  Possible values
for boundary are described in the page for
.BR QccWAVWavelet (3).
Not all boundary methods are permitted for
all wavelets; check the comments at the start of
.B FBK
or
.B LFT
files for possible values for particular wavelets.
The default boundary-extension method is "symmetric".
.LP
If the
.B \-pw
option is specified, then each wavelet coefficient is
perceptually weighted after the DWT is performed.
Perceptual weights are given by the file
.I pcpfile
.RB ( PCP
format);
this file must be found in the path given by the
environment variable
.BR QCCPACK_WAVELET_PATH .
The perceptual weights in
.I pcpfile
are the
reciprocals of the just-noticeable distortions (JND) for the subbands.
That is, perceptual weighting is accomplished by multiplying
each wavelet coefficient by the appropriate weight given in
.I pcpfile 
(see 
.BR QccWAVPerceptualWeights (3)
and Hontsch 
.IR "et al" .).
In the absense of the
.B \-pw
option, no perceptual weighting is used (the default).
.LP
If the
.B \-m
option is specified,
.B imgdwt
performs a shape-adaptive DWT (SA-DWT), in which case,
.I mask
gives the transparency mask indicating which pixels are to be transformed.
.I mask
must be a grayscale image
in which black in this image indicates transparency, white
indicates full opacity, and gray tones are degrees of
translucence
(see
.BR QccAlpha (3)).
If the
.B \-m
option is not given, the usual DWT is used.
See 
.BR QccWAVSubbandPyramidShapeAdaptiveDWT2D (3)
and Li and Li for more 
information on the SA-DWT.
Note that
.B imgdist (1)
can perform "shape-adaptive" distortion
calculations, i.e., distortion only over
non-transparent regions, as needed
for measuring performance of SA-DWT based operations.
.LP
Note:
.B sbpsplit
may be used to split a subband pyramid into separate subband files.
.LP
.B imgidwt
implements the corresponding inverse DWT or inverse SA-DWT.
Perceptual weighting (if any) is removed before the inverse transform is
performed. The transparency
.IR mask ,
if used,
should be the same as used by
.BR imgdwt .
The number of levels of inverse DWT to be performed is extracted
from the header of
.IR subband_pyramid .
If the 
.B \-l
option is specified,
.I last_subband
gives the last subband to include in the inverse DWT.  That is,
.B imgidwt
will zero out the subbands numbered greater than or equal to
.I last_subband
+ 1
in the subband pyramid before taking the inverse transform.
The subband zeroing is performed via a call to
.BR QccWAVSubbandPyramidZeroSubband (3).
For example, to see what retaining only the baseband coefficients
looks like, use
.I last_subband
= 0.
.SH "SEE ALSO"
.BR sbpsplit (1),
.BR sbpassemble (1),
.BR QccWAVSubbandPyramidDWT (3),
.BR QccWAVSubbandPyramidInverseDWT (3),
.BR QccWAVSubbandPyramidShapeAdaptiveDWT (3),
.BR QccWAVSubbandPyramidInverseShapeAdaptiveDWT (3),
.BR QccWAVSubbandPyramidZeroSubband (3),
.BR QccWAVPerceptualWeights (3),
.BR QccWAVSubbandPyramidZeroSubband (3),
.BR QccWAVSubbandPyramid (3),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

A. Cohen, I. Daubechies, J. C. Feauveau, "Biorthogonal Bases of
Compactly Supported Wavelets," 
.IR "Communications on Pure and Applied Mathematics" ,
vol. 45, no. 5, pp. 485-560, May 1992.

M. Antonini, M. Barlaud, P. Mathieu, I. Daubechies, "Image Coding Using
Wavelet Transform," 
.IR "IEEE Transactions on Image Processing" ,
vol. 1, no. 2, pp. 205-220, April 1992.

I. Hontsch, L. J. Karam, and R. J. Safranek,
"A Perceptually Tuned Embedded Zerotree Image Coder,"
.IR "Proceedings of the IEEE International Conference on Image Processing" ,
vol. 1, pp. 41-44, October 1997.

S. Li and W. Li, "Shape-Adaptive Discrete Wavelet Transforms for
Arbitrarily Shaped Visual Object Coding,"
.IR "IEEE Transactions on Circuits and Systems for Video Coding" ,
vol. 10, pp. 725-743, August 2000.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/imgdwt.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "imgdwt.h"

#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-pw %: %s:pcpfile] [-m %: %s:mask] [-nl %d:levels] %s:imgfile %s:subband_pyramid"

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

QccWAVPerceptualWeights PerceptualWeights;
int UsePerceptualWeights = 0;

int NumLevels = 1;
int NumSubbands;

QccIMGImage Mask;
int MaskSpecified = 0;
int MaskNumRows, MaskNumCols;
QccWAVSubbandPyramid MaskSubbandPyramid;

QccIMGImage InputImage;
QccWAVSubbandPyramid SubbandPyramid;

int ImageNumCols;
int ImageNumRows;


int main(int argc, char *argv[])
{
  
  QccInit(argc, argv);

  QccWAVWaveletInitialize(&Wavelet);
  QccWAVPerceptualWeightsInitialize(&PerceptualWeights);
  QccIMGImageInitialize(&Mask);
  QccIMGImageInitialize(&InputImage);
  QccWAVSubbandPyramidInitialize(&SubbandPyramid);
  QccWAVSubbandPyramidInitialize(&MaskSubbandPyramid);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         WaveletFilename,
                         Boundary,
                         &UsePerceptualWeights,
                         PerceptualWeights.filename,
                         &MaskSpecified,
                         Mask.filename,
                         &NumLevels,
                         InputImage.filename,
                         SubbandPyramid.filename))
    QccErrorExit();
      
  if (NumLevels < 1)
    {
      QccErrorAddMessage("%s: Number of levels must be 1 or greater",
                         argv[0]);
      QccErrorExit();
    }
  NumSubbands = QccWAVSubbandPyramidNumLevelsToNumSubbands(NumLevels);

  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (UsePerceptualWeights)
    {
      if (QccWAVPerceptualWeightsRead(&PerceptualWeights))
        {
          QccErrorAddMessage("%s: Error calling QccWAVPerceptualWeightsRead()",
                             argv[0]);
          QccErrorExit();
        }
      if (PerceptualWeights.num_subbands != NumSubbands)
        {
          QccErrorAddMessage("%s: Number of subbands in %s does not match specified number of levels (%d)",
                             argv[0],
                             PerceptualWeights.filename,
                             NumLevels);
          QccErrorExit();
        }
    }

  if (QccIMGImageRead(&InputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageColor(&InputImage))
    {
      QccErrorAddMessage("%s: Color images not supported",
                         argv[0]);
      QccErrorExit();
    }

  ImageNumCols = InputImage.Y.num_cols;
  ImageNumRows = InputImage.Y.num_rows;

  if (MaskSpecified)
    {
      if (QccIMGImageRead(&Mask))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageColor(&Mask))
        {
          QccErrorAddMessage("%s: Mask must be grayscale",
                             argv[0]);
          QccErrorExit();
        }
      if (QccIMGImageGetSize(&Mask,
                             &MaskNumRows, &MaskNumCols))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
                             argv[0]);
          QccErrorExit();
        }

      if ((MaskNumRows != ImageNumRows) ||
          (MaskNumCols != ImageNumCols))
        {
          QccErrorAddMessage("%s: Mask must be same size as image",
                             argv[0]);
          QccErrorExit();
        }
    }

  SubbandPyramid.num_levels = 0;
  SubbandPyramid.num_rows = ImageNumRows;
  SubbandPyramid.num_cols = ImageNumCols;
  SubbandPyramid.matrix = InputImage.Y.image;
  
  if (!MaskSpecified)
    {
      if (QccWAVSubbandPyramidDWT(&SubbandPyramid,
                                  NumLevels,
                                  &Wavelet))
        {
          QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidDWT()",
                             argv[0]);
          QccErrorExit();
        }
    }
  else
    {
      MaskSubbandPyramid.num_levels = 0;
      MaskSubbandPyramid.num_rows = ImageNumRows;
      MaskSubbandPyramid.num_cols = ImageNumCols;
      MaskSubbandPyramid.matrix = Mask.Y.image;
  
      if (QccWAVSubbandPyramidShapeAdaptiveDWT(&SubbandPyramid,
                                               &MaskSubbandPyramid,
                                               NumLevels,
                                               &Wavelet))
        {
          QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidShapeAdaptiveDWT()",
                             argv[0]);
          QccErrorExit();
        }
    }

  if (UsePerceptualWeights)
    if (QccWAVPerceptualWeightsApply(&SubbandPyramid,
                                     &PerceptualWeights))
      {
        QccErrorAddMessage("%s: Error calling QccWAVPerceptualWeightsApply()",
                           argv[0]);
        QccErrorExit();
      }

  if (QccWAVSubbandPyramidWrite(&SubbandPyramid))
    {
      QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidWrite()",
                         argv[0]);
      QccErrorExit();
    }

  QccWAVWaveletFree(&Wavelet);
  QccWAVPerceptualWeightsFree(&PerceptualWeights);
  QccIMGImageFree(&Mask);
  QccIMGImageFree(&InputImage);

  QccExit;
}
